Sources/Overlays/_Testing_Foundation/Attachments/Attachment+URL.swift

```diff
@@ -0,0 +1,161 @@
+      preferredName = {
+#if SWT_TARGET_OS_APPLE && canImport(UniformTypeIdentifiers)
+        if #available(_uttypesAPI, *), let tgz = UTType.tgz {
+          return (preferredName as NSString).appendingPathExtension(for: tgz)
+        }
+#endif
+        return (preferredName as NSString).appendingPathExtension("tgz") ?? preferredName
+      }()
+
+      try await self.init(Data(compressedContentsOfDirectoryAt: url), named: preferredName, sourceLocation: sourceLocation)
+    } else {
+      // Load the file.
+      try self.init(Data(contentsOf: url, options: [.mappedIfSafe]), named: preferredName, sourceLocation: sourceLocation)
+    }
+  }
+}
+
+// MARK: - Attaching directories
+
+extension Data {
+  /// Initialize an instance of this type by compressing the contents of a
+  /// directory.
+  ///
+  /// - Parameters:
+  ///   - directoryURL: A URL referring to the directory to attach.
+  ///
+  /// - Throws: Any error encountered trying to compress the directory, or if
+  ///   directories cannot be compressed on this platform.
+  ///
+  /// This initializer asynchronously compresses the contents of `directoryURL`
+  /// into an archive (currently of `.tgz` format, although this is subject to
+  /// change) and stores a mapped copy of that archive.
+  init(compressedContentsOfDirectoryAt directoryURL: URL) async throws {
+    let temporaryName = "\(UUID().uuidString).tgz"
+    let temporaryURL = FileManager.default.temporaryDirectory.appendingPathComponent(temporaryName)
+
+#if !SWT_NO_PROCESS_SPAWNING
+#if os(Windows)
+    let tarPath = #"C:\Windows\System32\tar.exe"#
+#else
+    let tarPath = "/usr/bin/tar"
+#endif
+    let sourcePath = directoryURL.fileSystemPath
+    let destinationPath = temporaryURL.fileSystemPath
+    defer {
+      try? FileManager().removeItem(at: temporaryURL)
+    }
+
+    try await withCheckedThrowingContinuation { continuation in
+      do {
+        _ = try Process.run(
+          URL(fileURLWithPath: tarPath, isDirectory: false),
+          arguments: ["--create", "--gzip", "--directory", sourcePath, "--file", destinationPath, "."]
+        ) { process in
+          let terminationReason = process.terminationReason
+          let terminationStatus = process.terminationStatus
+          if terminationReason == .exit && terminationStatus == EXIT_SUCCESS {
+            continuation.resume()
+          } else {
+            let error = CocoaError(.fileWriteUnknown, userInfo: [
+              NSLocalizedDescriptionKey: "The directory at '\(sourcePath)' could not be compressed.",
+            ])
+            continuation.resume(throwing: error)
+          }
+        }
+      } catch {
+        continuation.resume(throwing: error)
+      }
+    }
+
+    try self.init(contentsOf: temporaryURL, options: [.mappedIfSafe])
+#else
+    throw CocoaError(.featureUnsupported, userInfo: [NSLocalizedDescriptionKey: "This platform does not support attaching directories to tests."])
```

Comment:
Slightly pedantic, but it's only the creation of the attachment which is unsupported here, not _attaching_ it to a test, right?

```suggestion
    throw CocoaError(.featureUnsupported, userInfo: [NSLocalizedDescriptionKey: "This platform does not support creating attachments with the contents of a directory."])
```